File: python/wine_quality_analysis.py
"""
Wine Quality Prediction Using Machine Learning
------------------------------------------------
Predicts the sensory quality score of white "Vinho Verde" wine from 11
physicochemical measurements, comparing Multiple Linear Regression,
Random Forest, and XGBoost.

Dataset: UCI Wine Quality (white wine subset), Cortez et al. (2009).

Usage:
    python wine_quality_analysis.py
"""
import os
import urllib.request

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from xgboost import XGBRegressor

RANDOM_STATE = 42

DATA_URL = (
    "https://archive.ics.uci.edu/ml/machine-learning-databases/"
    "wine-quality/winequality-white.csv"
)
HERE = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(HERE, "..", "data")
DATA_PATH = os.path.join(DATA_DIR, "winequality-white.csv")
FIG_DIR = os.path.join(HERE, "..", "figures")
RESULTS_DIR = os.path.join(HERE, "..", "results")


def load_data() -> pd.DataFrame:
    """Load the dataset, downloading it from UCI if not already present locally."""
    os.makedirs(DATA_DIR, exist_ok=True)
    if not os.path.exists(DATA_PATH):
        print(f"Dataset not found locally, downloading from {DATA_URL} ...")
        urllib.request.urlretrieve(DATA_URL, DATA_PATH)
    df = pd.read_csv(DATA_PATH, sep=";")
    return df


def explore(df: pd.DataFrame):
    os.makedirs(FIG_DIR, exist_ok=True)

    print("\n=== Dataset shape ===")
    print(df.shape)

    print("\n=== Missing values ===")
    print(df.isna().sum().sum(), "missing values found")

    print("\n=== Descriptive statistics ===")
    print(df.describe().T.round(3))

    # Correlation heatmap
    plt.figure(figsize=(10, 8))
    corr = df.corr()
    sns.heatmap(corr, annot=False, cmap="RdBu", center=0, square=True)
    plt.title("Correlation Heatmap of Features")
    plt.tight_layout()
    plt.savefig(os.path.join(FIG_DIR, "correlation_heatmap.png"), dpi=150)
    plt.close()
    print(f"\nSaved correlation heatmap -> {FIG_DIR}/correlation_heatmap.png")

    print("\nCorrelation with quality (sorted):")
    print(corr["quality"].sort_values(ascending=False))


def preprocess(df: pd.DataFrame):
    X = df.drop(columns=["quality"])
    y = df["quality"]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.30, random_state=RANDOM_STATE
    )

    scaler = StandardScaler()
    X_train_scaled = pd.DataFrame(
        scaler.fit_transform(X_train), columns=X_train.columns, index=X_train.index
    )
    X_test_scaled = pd.DataFrame(
        scaler.transform(X_test), columns=X_test.columns, index=X_test.index
    )

    return X_train, X_test, X_train_scaled, X_test_scaled, y_train, y_test


def evaluate(name, y_test, y_pred):
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    mae = mean_absolute_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    print(f"\n{name}")
    print(f"  RMSE: {rmse:.4f}")
    print(f"  R-squared: {r2:.4f}")
    print(f"  MAE: {mae:.4f}")
    return {"Model": name, "RMSE": rmse, "R-squared": r2, "MAE": mae}


def plot_feature_importance(name, model, feature_names):
    importances = pd.Series(model.feature_importances_, index=feature_names)
    importances = importances.sort_values(ascending=True)

    plt.figure(figsize=(8, 6))
    importances.plot(kind="barh", color="#4C72B0")
    plt.title(f"{name} — Feature Importance")
    plt.xlabel("Importance")
    plt.tight_layout()
    fname = f"feature_importance_{name.lower().replace(' ', '_')}.png"
    plt.savefig(os.path.join(FIG_DIR, fname), dpi=150)
    plt.close()
    print(f"Saved {name} feature importance -> {FIG_DIR}/{fname}")


def plot_model_comparison(results_df: pd.DataFrame):
    metrics = ["RMSE", "R-squared", "MAE"]
    fig, axes = plt.subplots(1, 3, figsize=(14, 5))
    palette = {"Multiple Linear Regression": "#E07A5F",
               "Random Forest": "#3D9970",
               "XGBoost": "#4C72B0"}

    for ax, metric in zip(axes, metrics):
        colors = [palette.get(m, "#888") for m in results_df["Model"]]
        ax.bar(results_df["Model"], results_df[metric], color=colors)
        ax.set_title(metric)
        ax.set_xticklabels(results_df["Model"], rotation=20, ha="right")

    plt.suptitle("Model Performance Comparison")
    plt.tight_layout()
    plt.savefig(os.path.join(FIG_DIR, "model_comparison.png"), dpi=150)
    plt.close()
    print(f"\nSaved model comparison chart -> {FIG_DIR}/model_comparison.png")


def main():
    df = load_data()
    explore(df)

    X_train, X_test, X_train_s, X_test_s, y_train, y_test = preprocess(df)

    results = []

    # --- Multiple Linear Regression ---
    lr = LinearRegression()
    lr.fit(X_train_s, y_train)
    pred_lr = lr.predict(X_test_s)
    results.append(evaluate("Multiple Linear Regression", y_test, pred_lr))

    print("\nLinear Regression coefficients:")
    print(pd.Series(lr.coef_, index=X_train.columns).sort_values(ascending=False))

    # --- Random Forest ---
    rf = RandomForestRegressor(
        n_estimators=500, max_features=3, random_state=RANDOM_STATE, n_jobs=-1
    )
    rf.fit(X_train, y_train)  # tree ensembles don't need scaling
    pred_rf = rf.predict(X_test)
    results.append(evaluate("Random Forest", y_test, pred_rf))
    plot_feature_importance("Random Forest", rf, X_train.columns)

    # --- XGBoost ---
    xgb = XGBRegressor(
        max_depth=6, learning_rate=0.1, n_estimators=100,
        random_state=RANDOM_STATE, objective="reg:squarederror"
    )
    xgb.fit(X_train, y_train)
    pred_xgb = xgb.predict(X_test)
    results.append(evaluate("XGBoost", y_test, pred_xgb))
    plot_feature_importance("XGBoost", xgb, X_train.columns)

    # --- Compare & save ---
    results_df = pd.DataFrame(results).sort_values("RMSE")
    print("\n=== Final Comparison ===")
    print(results_df.to_string(index=False))

    os.makedirs(RESULTS_DIR, exist_ok=True)
    results_df.to_csv(os.path.join(RESULTS_DIR, "metrics_summary.csv"), index=False)
    print(f"\nSaved metrics -> {RESULTS_DIR}/metrics_summary.csv")

    plot_model_comparison(results_df)


if __name__ == "__main__":
    main()

File: R/wine_quality_analysis.R
# =============================================================
# Wine Quality Prediction Using Machine Learning
# -------------------------------------------------------------
# Predicts the sensory quality score of white "Vinho Verde" wine
# from 11 physicochemical measurements. Compares Multiple Linear
# Regression, Random Forest, and XGBoost.
#
# Dataset: UCI Wine Quality (white wine subset), Cortez et al. (2009)
#
# Usage:
#   Rscript wine_quality_analysis.R
# =============================================================

required_packages <- c("randomForest", "xgboost", "ggplot2", "corrplot", "caret")
new_packages <- required_packages[!(required_packages %in% installed.packages()[, "Package"])]
if (length(new_packages) > 0) install.packages(new_packages, repos = "https://cloud.r-project.org")

library(randomForest)
library(xgboost)
library(ggplot2)
library(corrplot)
library(caret)

set.seed(42)

# --- Paths ---
data_dir <- file.path("..", "data")
fig_dir <- file.path("..", "figures")
results_dir <- file.path("..", "results")
dir.create(data_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(fig_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(results_dir, showWarnings = FALSE, recursive = TRUE)

data_path <- file.path(data_dir, "winequality-white.csv")
data_url <- paste0(
  "https://archive.ics.uci.edu/ml/machine-learning-databases/",
  "wine-quality/winequality-white.csv"
)

# --- 1. Load data (download if needed) ---
if (!file.exists(data_path)) {
  message("Dataset not found locally, downloading from UCI...")
  download.file(data_url, destfile = data_path, mode = "wb")
}
wine <- read.csv(data_path, sep = ";")

cat("\n=== Dataset dimensions ===\n")
print(dim(wine))

cat("\n=== Missing values ===\n")
print(sum(is.na(wine)))

cat("\n=== Summary statistics ===\n")
print(summary(wine))

# --- 2. Correlation heatmap ---
png(file.path(fig_dir, "correlation_heatmap.png"), width = 900, height = 800)
corr_matrix <- cor(wine)
corrplot(corr_matrix, method = "color", type = "upper", tl.col = "black",
         tl.srt = 45, title = "Correlation Heatmap of Features", mar = c(0, 0, 2, 0))
dev.off()
cat("\nSaved correlation heatmap ->", file.path(fig_dir, "correlation_heatmap.png"), "\n")

cat("\nCorrelation with quality (sorted):\n")
print(sort(corr_matrix[, "quality"], decreasing = TRUE))

# --- 3. Train/test split (70/30) ---
train_index <- createDataPartition(wine$quality, p = 0.70, list = FALSE)
train <- wine[train_index, ]
test <- wine[-train_index, ]

# Z-score scaling for the linear model (fit on train, applied to test)
preproc <- preProcess(train[, -ncol(train)], method = c("center", "scale"))
train_scaled <- predict(preproc, train)
test_scaled <- predict(preproc, test)

rmse <- function(actual, pred) sqrt(mean((actual - pred)^2))
mae <- function(actual, pred) mean(abs(actual - pred))
r_squared <- function(actual, pred) {
  1 - sum((actual - pred)^2) / sum((actual - mean(actual))^2)
}

results <- data.frame(Model = character(), RMSE = double(), R_squared = double(), MAE = double())

# --- 4. Multiple Linear Regression ---
lm_model <- lm(quality ~ ., data = train_scaled)
cat("\n=== Multiple Linear Regression summary ===\n")
print(summary(lm_model))

pred_lm <- predict(lm_model, newdata = test_scaled)
results <- rbind(results, data.frame(
  Model = "Multiple Linear Regression",
  RMSE = rmse(test$quality, pred_lm),
  R_squared = r_squared(test$quality, pred_lm),
  MAE = mae(test$quality, pred_lm)
))

# --- 5. Random Forest ---
rf_model <- randomForest(quality ~ ., data = train, ntree = 500, importance = TRUE)
print(rf_model)

png(file.path(fig_dir, "feature_importance_random_forest.png"), width = 900, height = 600)
varImpPlot(rf_model, main = "Random Forest — Feature Importance")
dev.off()
cat("\nSaved RF feature importance ->", file.path(fig_dir, "feature_importance_random_forest.png"), "\n")

pred_rf <- predict(rf_model, newdata = test)
results <- rbind(results, data.frame(
  Model = "Random Forest",
  RMSE = rmse(test$quality, pred_rf),
  R_squared = r_squared(test$quality, pred_rf),
  MAE = mae(test$quality, pred_rf)
))

# --- 6. XGBoost ---
train_matrix <- xgb.DMatrix(data = as.matrix(train[, -ncol(train)]), label = train$quality)
test_matrix <- xgb.DMatrix(data = as.matrix(test[, -ncol(test)]), label = test$quality)

xgb_model <- xgb.train(
  params = list(objective = "reg:squarederror", max_depth = 6, eta = 0.1),
  data = train_matrix,
  nrounds = 100,
  watchlist = list(train = train_matrix, test = test_matrix),
  early_stopping_rounds = 10,
  verbose = 0
)

importance_matrix <- xgb.importance(model = xgb_model)
png(file.path(fig_dir, "feature_importance_xgboost.png"), width = 900, height = 600)
xgb.plot.importance(importance_matrix, main = "XGBoost — Feature Importance")
dev.off()
cat("\nSaved XGBoost feature importance ->", file.path(fig_dir, "feature_importance_xgboost.png"), "\n")

pred_xgb <- predict(xgb_model, test_matrix)
results <- rbind(results, data.frame(
  Model = "XGBoost",
  RMSE = rmse(test$quality, pred_xgb),
  R_squared = r_squared(test$quality, pred_xgb),
  MAE = mae(test$quality, pred_xgb)
))

# --- 7. Compare & save ---
results <- results[order(results$RMSE), ]
cat("\n=== Final Comparison ===\n")
print(results)

write.csv(results, file.path(results_dir, "metrics_summary.csv"), row.names = FALSE)
cat("\nSaved metrics ->", file.path(results_dir, "metrics_summary.csv"), "\n")

results_long <- reshape(
  results,
  varying = c("RMSE", "R_squared", "MAE"),
  v.names = "Value",
  timevar = "Metric",
  times = c("RMSE", "R_squared", "MAE"),
  direction = "long"
)

p <- ggplot(results_long, aes(x = Model, y = Value, fill = Model)) +
  geom_bar(stat = "identity") +
  facet_wrap(~Metric, scales = "free_y") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 20, hjust = 1), legend.position = "none") +
  labs(title = "Model Performance Comparison", x = NULL, y = NULL)

ggsave(file.path(fig_dir, "model_comparison.png"), plot = p, width = 10, height = 5, dpi = 150)
cat("\nSaved model comparison chart ->", file.path(fig_dir, "model_comparison.png"), "\n")
